packages/duckietown/include/duckietown/dtros/phase_timing.py

```diff
@@ -0,0 +1,138 @@
+import time
+import inspect
+from collections import namedtuple, defaultdict
+from contextlib import contextmanager
+
+from .constants import PHASE_TIMER_MAX_KEEP_SEC
+
+TimeRange = namedtuple('TimeRange', 'start end')
+SinglePhaseStatistics = namedtuple('TimeRange', 'num_events frequency avg_duration filename lines line_nums')
+
+class PhaseTimer:
+
+    def __init__(self):
+        self.phases = defaultdict(SinglePhaseTimer)
+        self.time_phase = self._time_phase_notrecording
+        self.is_recording = False
+
+    def reset(self):
+        self.__init__()
+
+    def stop_recording(self):
+        self.time_phase = self._time_phase_notrecording
+        self.reset()
+        self.is_recording = False
+
+    def start_recording(self):
+        self.time_phase = self._time_phase_recording
+        self.is_recording = True
+
+    @contextmanager
+    def _time_phase_recording(self, phase_name):
+
+        # get and save the lines and the source code in this timing segment if not done already
+        if not self.phases[phase_name].context_set:
+            try:
+                start_frame = inspect.currentframe()
+                start_called_from_frame_line = inspect.getouterframes(start_frame)[2][2]
+                called_from_file = inspect.getfile(inspect.getouterframes(start_frame)[2][0])
+            except Exception as e:
+                print("Error in extracting the source code of the timing context (start): %s" % str(e))
+
+        # actual timing code (yield hands back to the code in the context)
+        time_start = time.time()
+        yield
+        time_end = time.time()
+        self.phases[phase_name].add_measurement(start=time_start, end=time_end)
+
+        # get and save the lines and the source code in this timing segment if not done already
+        if not self.phases[phase_name].context_set:
+            try:
+                end_frame = inspect.currentframe()
+                end_called_from_frame_line = inspect.getouterframes(end_frame)[2][2]
+                code_lines, start_idx = inspect.getsourcelines(inspect.getouterframes(end_frame)[2][0])
+                code_lines = code_lines[start_called_from_frame_line-start_idx:end_called_from_frame_line-start_idx+1]
+                self.phases[phase_name].add_context(filename=called_from_file,
+                                                    lines=code_lines,
```

Comment:
the actual code_lines are not needed, we don't want to pay in transport for them, [start, end] will be enough for us to create a link to GitHub.

Reply:
They are not being transported. The message doesn't have them. Left it here, just in case if we need it later. We can just comment it if it bothers you.